fix(semantic): allow assigning to `eval` and `arguments` in ambient context

camc314 opened the pull request:
https://github.com/microsoft/typescript/blob/7f802bbca25634c6bffe316de75771274f1d9b1d/src/compiler/binder.ts#L2670

https://www.typescriptlang.org/play/?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXxADcoIAKASgC55CctgBueAemfgHkBpAKFElgQp02PPFgBzZAFsQqDAGcK1WvSasOPPtDhI0mXPkSkiJaqmkAjEDAA0YmJJlz5Zy9arxzUqzDVsuvODagnoi+OKkAN4mEHYS0rIKAL7U0cQQrt7WcQ4Jzpk+SR5ePn7wAKIwMDgwgfw6QvqiABakANoxOY6J8gC61G0l2Z5uML3Fo2WV1TBAA

camc314 commented:
* **#12208** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12208?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12208?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=12208).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented on crates/oxc_semantic/src/checker/javascript.rs:
```diff
@@ -108,7 +108,14 @@ fn invalid_let_declaration(x0: &str, span1: Span) -> OxcDiagnostic {
     let strict_mode = ctx.strict_mode();
     // It is a Diagnostic if the StringValue of a BindingIdentifier is "eval" or "arguments" within strict mode code.
-    if strict_mode && matches!(ident.name.as_str(), "eval" | "arguments") {
+    if strict_mode
+        && matches!(ident.name.as_str(), "eval" | "arguments")
+        && ctx
+            .nodes
+            .parent_kind(ctx.current_node_id)
+            .as_function()
+            .is_none_or(|function| !matches!(function.r#type, FunctionType::TSDeclareFunction))
+    {
```
I'd assumed `declare function f({eval});` would also be legal. But it seems not: [TS Playground](https://www.typescriptlang.org/play/?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVIAoQA3KCALnlWQFsAjEGAGnlgHNaRUMBnS6+owCU-WgxgBuAFChIsBCnTY88NgQDeJMi3aduPAL6VNpClTGMdMDjS69RgmAZHnH02dDhI0mXPgAWBADaWhBWNnY8ALqUQQLiLPGMUS5JklJAA)

It might be worth making a comment to that effect to explain why only checking the parent and not searching further up ancestry.

overlookmotel replied:
Oh wait! This PR relates to `declare function eval();` not `declare function f(eval);`. Presumably we handle that one too.

Dunqing commented:
Oh, the check gets more complicated now! I will take a look tomorrow.

overlookmotel commented on crates/oxc_semantic/src/checker/javascript.rs:
```diff
@@ -107,9 +107,40 @@ fn invalid_let_declaration(x0: &str, span1: Span) -> OxcDiagnostic {
+                .is_some_and(|func| matches!(func.r#type, FunctionType::TSDeclareFunction))
+        };
+
+        let is_ok = match ctx.nodes.parent_kind(ctx.current_node_id) {
```
It'd be more performant to avoid repeatedly iterating down ancestors from current node.

```rs
let is_declare_function = |kind| {
    kind.as_function()
        .is_some_and(|func| matches!(func.r#type, FunctionType::TSDeclareFunction))
};

let parent_node_id = ctx.nodes.parent_id(ctx.current_node_id);
match ctx.nodes.kind(parent_node_id) {
    AstKind::FormalParameter(_) => {
        let grandparent = ctx.nodes.parent_kind(parent_node_id);
        is_declare_function(grandparent)
    }
    // etc
}
```

overlookmotel commented on crates/oxc_semantic/src/checker/javascript.rs:
```diff
@@ -107,9 +107,37 @@ fn invalid_let_declaration(x0: &str, span1: Span) -> OxcDiagnostic {
+        let parent = ctx.nodes.parent_node(ctx.current_node_id);
+        let is_ok = match parent.kind() {
+            AstKind::Function(func) => matches!(func.r#type, FunctionType::TSDeclareFunction),
+            AstKind::FormalParameter(_) => is_declare_function(&ctx.nodes.parent_kind(parent.id())),
+            AstKind::BindingRestElement(_) => {
+                let grand_parent = ctx.nodes.parent_node(parent.id());
+                matches!(grand_parent.kind(), AstKind::FormalParameters(_))
+                    && is_declare_function(&ctx.nodes.parent_kind(grand_parent.id()))
+            }
```
Oh do shut up you numpty.

overlookmotel commented:
### Merge activity

* **Jul 14, 7:14 PM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

camc314 committed:
fix(semantic): allow assigning to `eval` and `arguments` in ambient context (#12208)

https://github.com/microsoft/typescript/blob/7f802bbca25634c6bffe316de75771274f1d9b1d/src/compiler/binder.ts#L2670

https://www.typescriptlang.org/play/?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXxADcoIAKASgC55CctgBueAemfgHkBpAKFElgQp02PPFgBzZAFsQqDAGcK1WvSasOPPtDhI0mXPkSkiJaqmkAjEDAA0YmJJlz5Zy9arxzUqzDVsuvODagnoi+OKkAN4mEHYS0rIKAL7U0cQQrt7WcQ4Jzpk+SR5ePn7wAKIwMDgwgfw6QvqiABakANoxOY6J8gC61G0l2Z5uML3Fo2WV1TBAA
